app/src/main/java/com/example/moviesdbproject/rest/ApiRepositoryBase.kt

```diff
@@ -0,0 +1,40 @@
+package com.example.moviesdbproject.rest
+
+import com.example.moviesdbproject.exceptions.EmptyMoviesResponseException
+import com.example.moviesdbproject.exceptions.MoviesRequestErrorException
+import com.example.moviesdbproject.viewmodel.MovieState
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import retrofit2.Response
+
+/**
+ * Base api repository class
+ */
+open class ApiRepositoryBase {
+
+    private val _movieDetails: MutableStateFlow<MovieState> =
+        MutableStateFlow(MovieState.LOADING)
+
+    val movieDetails: StateFlow<MovieState>
+        get() = _movieDetails
+
+    /**
+     * executes network requests
+     */
+    protected suspend fun <T> executeRequest(
```

Comment:
is there anyway to generalize this to any sealed class (i.e. remove MovieState for some generic class which contains a SUCCESS or ERROR class inside)

Reply:
sealed interface ResponseState {} you want it like this??? give us an example please